src/object.d

```diff
@@ -4956,3 +4956,315 @@ string _d_assert_fail(string comp, A, B)(A a, B b) @nogc @safe nothrow pure
+    }
+    else static if (is(T U : U[]) && hasElaborateCopyConstructor!U)
+    {
+        // only do a postblit if the `U` requires it.
+        foreach (ref el; t)
+            __doPostblit(el);
+    }
+}
+
+T[] _d_arrayappendTTrace(T)(string file, int line, string funcname, ref T[] x, scope T[] y) @trusted pure
+{
+    import rt.profilegc : accumulate;
+
+    string typename = T.stringof;
+    version (tracegc)
+    {
+        printf("_d_arrayappendT type = %.*s x = %llu y = %llu file = '%.*s' line = %d function = '%.*s'\n",
+            cast(int)typename.length, typename.ptr,
+            cast(ulong)x.length,
+            cast(ulong)y.length,
+            file.length, file.ptr,
+            line,
+            funcname.length, funcname.ptr
+            );
+    }
+    accumulate(file, line, funcname, typename, T.sizeof * y.length);
+    return _d_arrayappendT(x, y);
+}
+
+/**
+ * Append y[] to array x[]
+ */
+T[] _d_arrayappendT_(T)(ref T[] x, scope T[] y) @trusted
+{
+    import core.stdc.string;
+    import core.internal.traits : Unqual;
+    alias unqualT = Unqual!T;
+
+    auto length = x.length;
+    auto sizeelem = T.sizeof;              // array element size
+    _d_arrayappendcTX(x, y.length);
+    memcpy(cast(Unqual!T *)&x[length], cast(Unqual!T *)&y[0], y.length * sizeelem);
+
+    // do postblit
+    __doPostblit(cast(Unqual!T[])x[length .. length + y.length]);
+    return x;
+}
+
+T[] _d_arrayappendcTXTrace(T)(string file, int line, string funcname, ref T[] px, size_t n) @trusted pure
+{
+    import rt.profilegc : accumulate;
+
+    string typename = T.stringof;
+    version (tracegc)
+    {
+        printf("_d_arrayappendcTX type = %.*s x = %llu n = %llu file = '%.*s' line = %d function = '%.*s'\n",
+            cast(int)typename.length, typename.ptr,
+            cast(ulong)px.length,
+            cast(ulong)n,
+            file.length, file.ptr,
+            line,
+            funcname.length, funcname.ptr
+            );
+    }
+    accumulate(file, line, funcname, typename, T.sizeof * n);
+    return _d_arrayappendcTX(px, n);
+}
+
+/**************************************
+ * Extend an array by n elements.
+ * Caller must initialize those elements.
+ */
+T[] _d_arrayappendcTX_(T)(ref T[] px, size_t n) @trusted
+{
+    import core.stdc.string;
+    import core.internal.traits : Unqual;
+    import core.memory : GC;
+    alias BlkInfo = GC.BlkInfo;
+    alias BlkAttr = GC.BlkAttr;
+
+    // FIXME: Access these directly from rt.lifetime
+    enum size_t PAGESIZE = 4096;
+    enum size_t LARGEPREFIX = 16;
+    enum size_t LARGEPAD = LARGEPREFIX + 1;
+
+    // Can not use `externDFunc` as that will cause a template instantiation loop.
+    template externDFuncMangle(string mangle, string func)
+    {
+        pragma(mangle, mangle) mixin(func);
+    }
+
+    alias __arrayStart = externDFuncMangle!("_D2rt8lifetime12__arrayStartFNaNbS4core6memory8BlkInfo_ZPv", "extern(D) void * externDFuncMangle(BlkInfo info) nothrow pure;");
```

Comment:
Looks like we have some structural problems in druntime, if this is required.  I don't know the solution at this time.

Reply:
Can a lot of this logic be encapsulated in a function in `rt.lifetime` so most of this can be replaced with a single function call, and avoid having to expose so many internals of `rt.lifetime` to  here?

Reply:
One thing I could do is that I could call `rt.lifetime._d_arrayappendcTX` directly instead of porting it. As I'm not sure I have time to start refactoring that function for this 2 week-block.

Reply:
Maybe that's best for now. If the compiler instantiates the template (wrapper), we can always port the implementation later on.